### ISSUE TITLE
zombies need better AI

### ISSUE THREAD

**Issue by jaxjaxk** [Zet0rz/nzombies#171]
the version i use is roughly a week old so it is possible this was fixed. the zombies are not very good. the issues i am having are as follows
1) zombies will get stuck behind doors and not go around.
2) zombies will not attack players if they are above them. ( this includes on a tall table or on top of the zombies)
3) the zombie health curve is wrong. same goes with the spawn number.

**Comment by minin43** [comment]
If anything, I'm annoyed that the zombies don't spawn far away from you, they ONLY spawn in areas right next to you, which means it's very easy to find a corner in a room and camp there.

**Comment by jaxjaxk** [comment]
make more spawns then mate..... their are 100 things you can do to fix that. I'm more annoyed that they get stuck in doors and other things. might be better for the doors to be removed when bought rather than open.

**Comment by Zet0rz** [comment]
1) That is a general Garry's Mod NextBot Nav problem, and there's nothing we can do to fix it other than to change their target position so they walk to another place, then to the door. The only way we can do that is to guess this place, and it would be really inconsistent. A possibility could be to use the Special Spawnpoints as checkpoints and walk to the nearest open one. They should be in the middle of any room. 

2) That is to do with their attack range and the fact that their actual position is calculated from their feet. I think it can be changed though. 

3) I am highly aware of the real curves and the fact that we don't use them, but we have to remember that Garry's Mod guns have different stats so we can never use the same curves and call it balanced.

**Comment by Zet0rz** [comment]
@minin43 I am aware of this and I also know why. Currently only the spawnpoints closer to you than the average distance to all spawnpoints are active. I will change it from `dist <= totaldist/totalspawnpoints` to something more like `dist <= (totaldist/totalspawnpoints)*0.75 + x` where X is a static number that should counteract the reduction on good map sizes, but make it bigger on maps smaller, and make it smaller on maps that are bigger.

**Comment by lolleko** [comment]
If you have problems with zombies getting stuck, make sure you put more effort into navmesh editing, it is so important to have a good navmesh! If you use an auto generated one and only edited it a bit or not at all the navigation will be bad, the auto generation is really awful. You are probably better of with flood generating larger areas and editing the rest youself.

tl;dr edit your navmeshes! (better NM  = better zombie behaviour), we cant improve the AI much more, in gmods current state.

**Comment by func-octo** [comment]
I have noticed that in kosovos, the zombies that spawn near the pistol wallbuy will try to get through the other barricade, up the stairs, and through that barricade instead of just using the other stairs, also 10/10 times the zombies get stuck on that corner in some sort of zen pose...

**Comment by Zet0rz** [comment]
> zen pose...

That's their jump animation, if they get stuck mid-jump, they go like that. It's quite funny to be honest, and they should respawn after a short time anyway, so it's fine :P
That being said, it is a navmesh issue and it should definitely be edited.

**Comment by lolleko** [comment]
New navmesh added in #175 it should fix most of the described issues.

**Comment by jaxjaxk** [comment]
try the AI from this addon. it works much much better

**Comment by jaxjaxk** [comment]
https://steamcommunity.com/sharedfiles/filedetails/?id=314667163

**Comment by jaxjaxk** [comment]
just do not use the whole addon, the models cause crashes in clients

**Comment by func-octo** [comment]
@Zet0rz  I've actually noticed that the stuck respawn only works about a third of the time.

**Comment by lolleko** [comment]
@jaxjaxk Uhm I am not sure what exactly is better in that AI? As far as i can tell there is nothing special or better about it.

**Comment by jaxjaxk** [comment]
well here is what it does better.
1) invincible zombie error is NOT a problem.
2) they don't get stuck in doors and actually know how to get around a navmesh
3) they attack props that are in their way.

**Comment by jaxjaxk** [comment]
i should also point out that nexbots have a handle stuck function in them that the current AI does not use.
the AI needs fixing because a better AI from the addon works better on the maps i use than the AI you use.

**Comment by lolleko** [comment]
I looked a t the source code of that "AI" it literally doesn't do anything else than we do in terms of pathing/navigating. The HandleStuck hook isn't used in that Haddon either and we dont use it because we have our own mechanics for this. What do you mean by invincible zombie error? And our zombies push away props and don't attack them...

EDIT: The only thing i noticed that is different is that the addon you pointed out uses smaller collision bounds for the zombies, this will make them get less stuck and overall navigate a bit better. But a really big downside is that bullet detection doesn't work for the outer parts of the body (arms, shoulders....).

EDIT2: And seriously about the doors, doors in this gamemode(cod) usually have a buy option and after you bought, them zombies won't get stuck behind them (because they are not there anymore). Also with proper navediting they won't get stuck either. You should really try to improve your map config and Navmesh!

**Comment by lolleko** [comment]
@ScriptnLua 

> I've actually noticed that the stuck respawn only works about a third of the time.

What do you mean? It works a 100% of the times if a zombie is stuck but it usually doesn't despawn them instant to make sure they are really stuck.

**Comment by func-octo** [comment]
So the zombies that get stuck in a zen pose (jump/fall animation) don't despawn because the gamemode thinks they are just falling. I would check for speed while falling to see if it hits zero.

**Comment by lolleko** [comment]
Could you make a screenshot of that zenpose? never saw that. Also zombies (should) despawn if stuck even while falling.

**Comment by Zet0rz** [comment]
The zen pose is their jumping animation frozen as they are stuck mid-jump. Also speed and whether it thinks it's falling should not matter, it checks if its position is close enough to the same one to count if it's stuck. And I have seen them respawn from this, it shouldn't be a problem.

> 1) invincible zombie error is NOT a problem.

Let us know what causes this bug so we can look into fixing it :)

> 2) they don't get stuck in doors and actually know how to get around a navmesh

Do you mean rotating doors that they get stuck on once opened? There's sadly not much to do about that, but you could place a prop/invisible wall to stop them from going there. If you mean prop doors (that you place down with Q), then _obviously_ they don't know about it, you'll need to use the Navlock tool to give them awareness. And you can't modify how Nextbots navigate navmeshes, so it's literally 100% the same.

> 3) they attack props that are in their way.

Ours push props, which is just as good. We shouldn't make them attack them, as then they'd break down doors at no cost to the players.

**Comment by func-octo** [comment]
> > 3) they attack props that are in the way.
> > Ours push props, which is just as good. We shouldn't make them attack them, as then they'd break down doors at no cost to the players.

Lol so true.

**Comment by jaxjaxk** [comment]
@Zet0rz 
1) invincible zombie error i suspect could be caused by several things. what i know is when they are bugged making them stuck makes them despawn as usual and when they attack players they do not do damage. this makes me belive that one of the following functions is to blame. A.self:IsValidTarget( self:GetTarget() ) B.self:TargetInRange( self:GetAttackRange() + 10 ). that or data.attackseq.dmgtimes = 0. it is posible that data.attackseq.dmgtimes is larger than 1 and the player is no longer valid because they are dead. i'm just throwing out ideas here

2) mainly i mean they will get over railings that they should be able to get over. right now they just walk up against the railings and stand there trying to move forward. for the most part i would love it if their was an option to have the doors get removed rather than opened with the AI getting stuck in them all the time it does not make the game mode look good.

3) have no idea what kind of idiot would put zombies on the other side of the door.... so you are talking about a situation that will only happen if an idiot makes the map. thing is with a lot of maps zombies will get stuck in props.

further more their is something i would like to ask.
the invisible wall maker for zombie spawn. are zombies not supposed to walk through those? it kind of seems like you made 2 tools for the same reason if that is true. i modified the entity so that zombies can go through but players can not, this keeps players from entering zombie spawn areas.

**Comment by Zet0rz** [comment]
1) I think you're encountering the flaming zombies. A bug causes their flames to mostly not appear. They do almost no damage but rather explodes after a short time of attacking. 

2) Railings? Doors? They're two different things. For railings, you should make the Navmesh so they don't even try to walk through it, or let them know they have to jump by marking it as a jump mesh. For rotating doors, you can use props to block them from getting behind it at all. 

3) If you make a full circular map and open all doors in one direction except the last one, zombies will spawn on the other side of this last one since that is the starting room. This becomes the case for _all_ non-linear maps that don't use the double-door method. As to getting stuck in props, do you mean they don't push enough? Haven't tested around much with that myself, we could calibrate it better. 

(4) I haven't messed around to much with the invisible doors, but it is planned to make them only block players if possible.

**Comment by jaxjaxk** [comment]
@i managed to get i fix for the invisible doors. it is not a 100% fix as players can get through if they try really hard, a zombie is walking through it and they get that small bit of luck. so basically you can break it if you intend to break it. in the end it is good enough for me because a player that wants to get past an invisible wall into a pit of death.... i don't think those exist.
